### ISSUE TITLE
Agent thinks I have more than one UHK connected to my computer

### ISSUE THREAD

**Issue by salamad** [UltimateHackingKeyboard/agent#1971]
Hello,

At the request of Nora, I am submitting my support request via this channel.

I received my UHK and was able to successfully install Agent on my Mac and update the firmware to the latest release. Everything has been working very well. However, I wanted to go in to Agent today to reprogram some keys and now Agent is saying "Multiple UHKs were found. Multiple devices aren't supported yet, so please connect only a single device to proceed further".

I only have a single UHK. I tried removing the additional key cluster and trackball modules but still continue to get the same error message.

I have already tried disconnecting the UHK and rebooting the computer but the problem still persists.

Any thoughts would be greatly appreciated.

**Comment by salamad** [comment]
Attached some screenshots for additional information.

![Screenshot 2023-02-24 at 3 05 38 PM](https://user-images.githubusercontent.com/456550/221280148-6f951f8d-9942-4370-8417-cb3ef0a8723c.png)
![Screenshot 2023-02-24 at 3 05 16 PM](https://user-images.githubusercontent.com/456550/221280150-3c46f6c3-cc9f-4aa1-abd3-daaa49c44a51.png)
![Screenshot 2023-02-24 at 3 05 00 PM](https://user-images.githubusercontent.com/456550/221280151-20105f8d-ae5a-4818-9786-15d51d7fac42.png)

**Comment by mondalaci** [comment]
Hi there, @salamad!

@ert78gb Any ideas?

**Comment by ert78gb** [comment]
Hi,

Could you attach the log of Agents from `~/Library/Logs/uhk-agent` folder?

**Comment by salamad** [comment]
Attached is the log file.

Thank you

[uhk-agent.log](https://github.com/UltimateHackingKeyboard/agent/files/10828536/uhk-agent.log)

**Comment by ert78gb** [comment]
Looks like Agent would like to connect to an Apple HID device. That make no sense. Please restart your machine. this is my first suggestion.

Do you use any software that track USB/HID devices?

**Comment by salamad** [comment]
I have restarted my machine 3 times. I have fully shutdown, disconnected all wires, and rebooted again and still the same problem.
I do not have any software to track USB/HID devices. At least there have been no changes to my computer since I was first able to use Agent and update the firmware.

**Comment by mondalaci** [comment]
@salamad Please directly connect your UHK to the computer (no KVMs, docking stations, USB hubs) and disconnect every other USB device.

**Comment by salamad** [comment]
@mondalaci I tried that as well. That should be the last entry/attempt in the log.

**Comment by ert78gb** [comment]
@salamad I created a new Agent [version](https://drive.google.com/file/d/1Tp_5z-5AkE29TZYrOw22Ra2OGpf5AvtA/view?usp=sharing) that contains the new `--print-usb-devices` command line argument. 
Please install the new version and run the `/Applications/UHK\ Agent.app/Contents/MacOS/UHK\ Agent -- --print-usb-devices` command in terminal. It will list every usb devices to the standard output. Please share the output with us. thx

**Comment by salamad** [comment]
@ert78gb please see output of command attached.

BTW, something happened last night that had never happened before. In the middle of me working, the left side of the keyboard completely died and was unresponsive. I disconnected and reconnected the cable that connects the halves and it started working again. This morning, when the keyboard was idle, the LED display started flickering several times, but it didn't fail. It just flickered. Just FYI.

Thank you

[usb_devices.txt](https://github.com/UltimateHackingKeyboard/agent/files/10831959/usb_devices.txt)

**Comment by mondalaci** [comment]
@salamad As for the flicker issue please see https://ultimatehackingkeyboard.com/knowledgebase?q=flicker and if it persists, email us to support at ultimatehackingkeyboard dot com

**Comment by ert78gb** [comment]
@salamad do you use any virtualisation software like WmWare of VirtualBox etc?

**Comment by salamad** [comment]
I don't use any virtualization software. However, I do have VirtualBox installed but have never used it though and don't have any guest OS setup either.

**Comment by ert78gb** [comment]
The problem is UHK keyboard has 4 USB device and and every device has an individual interface id, but in you case all of the 4 devices visible as 0 and it cause issues. I have never experienced it earlier. Could you try UHK and Agent communication on other computer? If it works on other computer could you uninstall the VirtualBox just to test is it cause the issue or not? Or update VirtualBox to the latest version. 
We experienced earlier these softwares cause issues

**Comment by salamad** [comment]
Hi @ert78gb 

I do have another mac but that's my work computer and it's fully locked. It won't let me install Agent. Therefore, I uninstalled VirtualBox using their uninstall tool and restarted my computer. Agent still gives me the same error. Attached is the updated output from the --print-usb-devices command.

[usb_devices.txt](https://github.com/UltimateHackingKeyboard/agent/files/10832738/usb_devices.txt)

Thank you

**Comment by salamad** [comment]
I also purchased a license of CleanMyMac X to uninstall any non-essential apps, extensions, etc and update any software that was identified as being out of date. Unfortunately, that did not help either and I still have the same problem.

**Comment by ert78gb** [comment]
What is your exact MacOS version number?

I use 13.2.1 and I don't experience any issue

**Comment by ert78gb** [comment]
If I see well well in the logs, you use newer macos version. Nodejs identifies your OS as `22.4.0`. 
Nodejs returns with `22.3.0` for  MacOs Ventura 13.2.1.
Based on the https://www.macrumors.com/2023/02/17/macos-ventura-13-3-public-beta-1/ article `22.4.0` is the next minor version of MacOs.
I don't know this every interface value = 0 is a MacOS bug or it will the behaviour of the next MacOS release, but my assumption is it is a bug in the OS, because the `path` property of the device information contains the proper interface version

**Comment by salamad** [comment]
That could be the problem. I do run macOS Ventura beta 13.3 (22E5219e). Just wondering why would it work 2 weeks ago and not now any more.

**Comment by ert78gb** [comment]
Based on the logs that you sent, 2 weeks ago you used earlier version of the OS. 

```
[2023-02-15 11:50:41.026] [info]  [AppService] getAppStartInfo response: {
  commandLineArgs: { modules: false, 'disable-agent-update-protection': false },
  platform: 'darwin',
  osVersion: '22.3.0',
  udevFileContent: ''
}
```

```
[2023-02-23 09:33:30.574] [info]  [AppService] getAppStartInfo response: {
  commandLineArgs: { modules: false, 'disable-agent-update-protection': false },
  platform: 'darwin',
  osVersion: '22.4.0',
  udevFileContent: ''
}
```

I don't see any Agent app started log entry between [2023-02-15 11:50:41.026] and [2023-02-23 09:33:30.574], so looks like there was the OS upgrade between these dates

**Comment by salamad** [comment]
Ok. So it's possible MacOS upgraded in between since I have automatic updates enabled.

**Comment by salamad** [comment]
I will need to see what happens after the next OS upgrade if the issue gets resolved or, if this is the future expected behavior, a new Agent version that supports that... unless of course you have any other ideas.

I can try taking the keyboard to someone else's Mac to reprogram the keys I need to change and then bring that back. I believe the settings will be persisted in the keyboard itself correct?

Thank you

**Comment by salamad** [comment]
I could also report the bug to Apple beta program if you believe this is incorrect behavior on their part.

**Comment by ert78gb** [comment]
My assumption is that it is an OS bug, because every OS fills the `interface` property of the USB device. We know Apple thinks different so we will see. Of course if it is the behaviour of the new MacOs we will fix it ASAP.

We highly appreciate if you report it to Apple and keep us in the loop. I don't report it because if they have further questions or have to test something on the new OS I can't do it. 

Yes, the settings stored on the keyboard.

**Comment by salamad** [comment]
Thank you for all your support. I'll report it to Apple and keep you informed.

**Comment by colinjoy** [comment]
Seeing the same problem since updating to macOS 13.3

System Version:	macOS 13.3 (22E252)
Kernel Version:	Darwin 22.4.0

**Comment by ert78gb** [comment]
I haven't updated yet to this version. @salamad Did you reported it and got any response?

**Comment by NSHenry** [comment]
Same issue here. 
macOS 13.3 (22E252)

**Comment by enrico2828** [comment]
I just received my brand new, 2nd UHK and was ready to set it up, but I hit the same issue. macOS 13.3 ...

**Comment by mondalaci** [comment]
@ert78gb will fix this issue in several days.

**Comment by edward-aslin** [comment]
How do we implement this fix? Are you going to create a new release or do we have to build from source?

**Comment by ert78gb** [comment]
We will deploy a new version today or tomorrow morning.

**Comment by Homseng** [comment]
Same issue here: Multiple UHKs were found.
Ventura 13.3

**Comment by eric-anderson-ibotta** [comment]
What release/version will this be part of? Or will it require building the agent from source?

**Comment by mondalaci** [comment]
The newly released Agent 2.1.2 should solve this issue.
https://github.com/UltimateHackingKeyboard/agent/releases/latest
